lineapy/graph_reader/artifact_collection.py

```diff
@@ -42,55 +46,60 @@ def __init__(
         self.db: RelationalLineaDB = db
 
         self.input_parameters = input_parameters
+
+        # ------ Validate inputs ------
+        # check if any duplicates exist in input parameters
         if len(input_parameters) != len(set(input_parameters)):
             raise ValueError(
                 f"Duplicated input parameters detected in {input_parameters}"
             )
-
-        # Retrieve target artifact objects and group them by session ID
-        self.target_artifacts_by_session = (
-            self._get_artifacts_grouped_by_session(target_artifacts)
-        )
-
-        # Retrieve reuse precomputed artifact objects and group them by session ID
-        self.pre_computed_artifacts_by_session = (
-            self._get_artifacts_grouped_by_session(
-                reuse_pre_computed_artifacts
-            )
+        # Check if target artifacts name has been repeated in the input list
+        check_duplicates(target_artifacts)
+        # Check if entries in reuse pre computed artifacts list have been repeated
+        check_duplicates(reuse_pre_computed_artifacts)
+
+        # ------ Initialize Linea Artifacts from artifactdef ------
+        # Retrieve target artifact objects
+        self.target_artifacts: List[
+            LineaArtifact
+        ] = get_db_artifacts_from_artifactdef(self.db, target_artifacts)
+
+        # Retrieve reuse precomputed artifact objects
+        self.pre_computed_artifacts: List[
+            LineaArtifact
+        ] = get_db_artifacts_from_artifactdef(
+            self.db, reuse_pre_computed_artifacts
         )
 
+        # ------ Group artifacts by session using SessionArtifact ------
         # For each session, construct SessionArtifacts object
         self.session_artifacts: Dict[LineaID, SessionArtifacts] = {}
         for (
             session_id,
-            session_artifacts,
-        ) in self.target_artifacts_by_session.items():
-            # TODO: LIN-653, LIN-640 only collect matched pre_computed artifacts in this Session
-            # instead of grabbing all Artifacts as currently required by SessionArtifact
-            pre_calculated_artifacts: List[LineaArtifact] = sum(
-                self.pre_computed_artifacts_by_session.values(), []
-            )
+            target_artifacts_by_session,
+        ) in get_artifacts_grouped_by_session(self.target_artifacts):
```

Comment:
This is the only place where grouping by session id is truely needed. SessionArtifacts is the dataclass to hold this grouping. Knowing that, there should not be a need to also retain another var that groups essentially the same way SessionArtifacts does. SessionArtifacts strictly provides more features over any dict-based structure and hence the old dict was refactored into a generator.